### PR TITLE
fix: restart onSnapshot() listeners that stop receiving updates

### LINKED ISSUE #1057: Collection group query onSnapshot stops being called
#### Environment details

  - OS: Alpine
  - Node.js version: 12.16
  - npm version: 
  - `@google-cloud/firestore` version: 3.7.5

#### Steps to reproduce

We have a gateway that listens to a collection group and processes new snapshots (frequency: 5 - 100 snapshots / minute):
```
  var eventQuery = database.collectionGroup('events')
    .where('state', '==', messageState.pending)
    .orderBy('createdAt');

  eventQuery.onSnapshot(processEventSnapshot);
```

Lately this stops working after some time and we stop getting new snapshots. (about a pair of hours).
We don't know what causes this. Obviously there are new events that should trigger onSnapshot since restarting the process solves the problem.

Here are the logs of our gateway at 5 am:
```
Firestore (3.7.5) 2020-04-30T04:59:12.020Z ZOE6F [Firestore.requestStream]: Received response: {"documentChange": ..., "name":"projects/X/databases/(default)/documents/tenants/X/channels/315559751805571/streams/3427895583892103/events/JfQzblpW8JqsKetlT5w3","createTime":{"seconds":"1588222752","nanos":1119000},"updateTime":{"seconds":"1588222752","nanos":1119000}}},"responseType":"documentChange"}
Firestore (3.7.5) 2020-04-30T04:59:12.021Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,223,152,212,231,174,143,233,2]},"readTime":{"seconds":"1588222752","nanos":1119000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T04:59:12.021Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T04:59:12.021Z ZOE6F [Watch.pushSnapshot]: Sending snapshot with 1 changes and 3182 documents
Firestore (3.7.5) 2020-04-30T05:00:13.077Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,141,195,227,132,175,143,233,2]},"readTime":{"seconds":"1588222813","nanos":69709000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:00:13.077Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:01:11.017Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,162,246,179,160,175,143,233,2]},"readTime":{"seconds":"1588222871","nanos":10082000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:01:11.017Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:02:41.039Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,203,179,170,203,175,143,233,2]},"readTime":{"seconds":"1588222961","nanos":31627000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:02:41.039Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:04:11.062Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,226,247,160,246,175,143,233,2]},"readTime":{"seconds":"1588223051","nanos":54050000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:04:11.062Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:05:41.083Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,187,183,151,161,176,143,233,2]},"readTime":{"seconds":"1588223141","nanos":75899000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:05:41.083Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:07:11.105Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,233,249,141,204,176,143,233,2]},"readTime":{"seconds":"1588223231","nanos":98089000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:07:11.105Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:08:41.127Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,139,185,132,247,176,143,233,2]},"readTime":{"seconds":"1588223321","nanos":119883000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:08:41.127Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:10:11.149Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,225,251,250,161,177,143,233,2]},"readTime":{"seconds":"1588223411","nanos":142113000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:10:11.149Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:11:41.170Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,230,185,241,204,177,143,233,2]},"readTime":{"seconds":"1588223501","nanos":163750000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:11:41.170Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:13:11.193Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,128,254,231,247,177,143,233,2]},"readTime":{"seconds":"1588223591","nanos":186176000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:13:11.193Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:14:41.216Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,139,198,222,162,178,143,233,2]},"readTime":{"seconds":"1588223681","nanos":209099000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:14:41.216Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:16:11.239Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,172,137,213,205,178,143,233,2]},"readTime":{"seconds":"1588223771","nanos":231404000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:16:11.239Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:17:41.261Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,205,200,203,248,178,143,233,2]},"readTime":{"seconds":"1588223861","nanos":253197000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:17:41.261Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:19:11.282Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,215,136,194,163,179,143,233,2]},"readTime":{"seconds":"1588223951","nanos":275095000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:19:11.283Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:20:41.304Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,144,203,184,206,179,143,233,2]},"readTime":{"seconds":"1588224041","nanos":297296000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:20:41.304Z ZOE6F [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T05:22:11.326Z ZOE6F [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[],"targetChangeType":"NO_CHANGE","cause":null,"resumeToken":{"type":"Buffer","data":[10,9,8,168,138,175,249,179,143,233,2]},"readTime":{"seconds":"1588224131","nanos":319080000}},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T05:22:11.326Z ZOE6F [Watch.onData]: Processing target change
```

This goes on until 6:07am, when our gateway restarted (it restarts every 3 hours to have some tolerance to these types of failures) and normal activity is resumed...

```
Firestore (3.7.5) 2020-04-30T06:08:12.896Z ##### [Firestore]: Initialized Firestore
Firestore (3.7.5) 2020-04-30T06:08:12.905Z bHJ8x [ClientPool.acquire]: Creating a new client
Firestore (3.7.5) 2020-04-30T06:08:13.006Z ##### [Firestore]: Initialized Firestore GAPIC Client
Firestore (3.7.5) 2020-04-30T06:08:13.006Z bHJ8x [Firestore.requestStream]: Sending request: {"database":"projects/X/databases/(default)","addTarget":{"query":{"parent":"projects/X/databases/(default)/documents","structuredQuery":{"from":[{"collectionId":"events","allDescendants":true}],"where":{"fieldFilter":{"field":{"fieldPath":"state"},"op":"EQUAL","value":{"stringValue":"PENDING"}}},"orderBy":[{"field":{"fieldPath":"createdAt"},"direction":"ASCENDING"}]}},"targetId":1}}
Firestore (3.7.5) 2020-04-30T06:08:13.016Z bHJ8x [Firestore._initializeStream]: Sending request: {"database":"projects/X/databases/(default)","addTarget":{"query":{"parent":"projects/X/databases/(default)/documents","structuredQuery":{"from":[{"collectionId":"events","allDescendants":true}],"where":{"fieldFilter":{"field":{"fieldPath":"state"},"op":"EQUAL","value":{"stringValue":"PENDING"}}},"orderBy":[{"field":{"fieldPath":"createdAt"},"direction":"ASCENDING"}]}},"targetId":1}}
Firestore (3.7.5) 2020-04-30T06:08:13.028Z bHJ8x [Firestore._initializeStream]: Marking stream as healthy
Firestore (3.7.5) 2020-04-30T06:08:13.029Z bHJ8x [Firestore._initializeStream]: Releasing stream
Firestore (3.7.5) 2020-04-30T06:08:13.029Z bHJ8x [Watch.initStream]: Opened new stream
Firestore (3.7.5) 2020-04-30T06:08:13.295Z bHJ8x [Firestore.requestStream]: Received response: {"targetChange":{"targetIds":[1],"targetChangeType":"ADD","cause":null,"resumeToken":{"type":"Buffer","data":[]},"readTime":null},"responseType":"targetChange"}
Firestore (3.7.5) 2020-04-30T06:08:13.295Z bHJ8x [Watch.onData]: Processing target change
Firestore (3.7.5) 2020-04-30T06:08:13.372Z bHJ8x [Firestore.requestStream]: Received response: {"documentChange":... "name":"projects/X/databases/(default)/documents/tenants/X/channels/315559751805571/streams/2250381201726111/events/DP433Z2tOQsbUog3H7ZX","createTime":{"seconds":"1583919154","nanos":806271000},"updateTime":{"seconds":"1583919154","nanos":806271000}}},"responseType":"documentChange"}
Firestore (3.7.5) 2020-04-30T06:08:13.373Z bHJ8x [Watch.onData]: Processing change event
Firestore (3.7.5) 2020-04-30T06:08:13.373Z bHJ8x [Watch.onData]: Received document change
```

The strangest part is, we have 2 different firestore projects and each with a different instance of the same gateway code, and this only happens in one of them.

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
https://github.com/googleapis/nodejs-firestore/issues/1057 describes an issue that we haven't been able to resolve on the backend. It seems that for large Watch snapshots (1000+ documents), the backend sometime stops responding and fails to deliver any further Watch updates. This PR adds a 2 minute timeout that automatically restarts the Watch stream and resumes the Query from the last known state.

The 2 minute timeout is many times larger than the 30 second interval at which Watch is expected to send a target change even if no documents have changed. 

Fixes #1057

### CODE REVIEW

**Comment by chrisfraser** [comment]
Great news. This is killing us. @wilhuff there is a virtual hug in this for you if you prioritize this😂 🙏🏼

**Comment by wilhuff** [comment]
This change seems like it would be useful on the client SDKs as well. Are you intending to port this there?

**Comment by chrisfraser** [comment]
I have a suspicion that this is happening in the Android client sdk, but have not been able to catch it with logs yet.

File: dev/test/bulk-writer.ts [hunk 1]
```diff
@@ -550,7 +550,7 @@ describe('BulkWriter', () => {
   });
 
   describe('500/50/5 support', () => {
-    afterEach(() => setTimeoutHandler(setTimeout));
+    afterEach(() => resetTimeoutHandler());
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/test/bulk-writer.ts]: `resetTimeoutHandler()` changes this back to returning a properly functioning NodeJS.Timeout. We don't use this in our tests though, so I rolled it back.

File: dev/test/transaction.ts [hunk 2]
```diff
@@ -353,7 +351,7 @@ function runTransaction<T>(
       });
     } finally {
       expect(expectedRequests.length).to.equal(0);
-      setTimeoutHandler(defaultTimeoutHandler);
+      resetTimeoutHandler();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/test/transaction.ts]: Yes, it does (an AssertionFailure). Changed order.

**Comment by chrisfraser** [comment]
I can confirm that when things die it does not effect the whole stream. Just snapshot listeners on specific collections. I tried a workaround that had the services listen to a "pingOfDeath" collection and had a scheduled function update a timestamp there every minute. If the service did not receive a pin within 2 minutes, I nuked the server. We would still get the ping when other collection snapshots had failed.

**Comment by JrSchild** [comment]
@schmidt-sebastian Apologies for digging up an old issue. It appears the back-end still has this issue. Over the weekend we observed a `Watch.resetIdleTimeout` log (same behavior, a large collection, no updates). As you mentioned a few years back that this is a [workaround](https://github.com/googleapis/nodejs-firestore/issues/1057#issuecomment-644382318), is there any reason to properly fix this on the backend?

It would still degrade performance, although not permanent, it may be inconvenient so it would be nice to prevent this code-path from occurring entirely.

**Comment by schmidt-sebastian** [comment]
cc @ehsannas

**Comment by ehsannas** [comment]
PTAL @dconeybe

**Pull request merged**